File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.GoTo.cs
using System;
using System.Threading;

namespace Monitor_Goto
{
    class Program
    {
        private object obj = new object();

        public void Method1()
        {
            Monitor.Enter(obj); // Compliant

            goto Release;

        Release:
            Monitor.Exit(obj);
        }

        public void Method2()
        {
            Monitor.Enter(obj); // FN

            goto DoNotRelease;

        Release:
            Monitor.Exit(obj);

        DoNotRelease:
            return;
        }
    }
}

File: analyzers/tests/SonarAnalyzer.Test/TestCases/SymbolicExecution/Roslyn/LocksReleasedAllPaths.Monitor.GoTo.vb
Imports System.Threading

Namespace Monitor_Goto

    Class Program

        Private Obj As New Object

        Public Sub Method1()
            Monitor.Enter(Obj) ' Compliant
            GoTo Release
Release:
            Monitor.Exit(Obj)
        End Sub

        Public Sub Method2()
            Monitor.Enter(Obj) ' FN
            GoTo DoNotRelease
Release:
            Monitor.Exit(Obj)
DoNotRelease:
        End Sub

    End Class

End Namespace
